IPython/terminal/shortcuts/auto_suggest.py

```diff
@@ -0,0 +1,287 @@
+import re
+import tokenize
+from io import StringIO
+from typing import Callable, List, Optional, Union
+
+from prompt_toolkit.buffer import Buffer
+from prompt_toolkit.key_binding import KeyPressEvent
+from prompt_toolkit.key_binding.bindings import named_commands as nc
+from prompt_toolkit.auto_suggest import AutoSuggestFromHistory, Suggestion
+from prompt_toolkit.document import Document
+from prompt_toolkit.history import History
+from prompt_toolkit.shortcuts import PromptSession
+
+from IPython.utils.tokenutil import generate_tokens
+
+
+def _get_query(document: Document):
+    return document.text.rsplit("\n", 1)[-1]
+
+
+class NavigableAutoSuggestFromHistory(AutoSuggestFromHistory):
+    """ """
+
+    def __init__(
+        self,
+    ):
+        self.skip_lines = 0
+        self._connected_apps = []
```

Comment:
Just coding defensively to ensure we always remove the handler even if we reconnect it from a different prompt session. I don't think this is likely to happen in core IPython today, but I did not feel confident that it will be true in the future and in third-party uses. I was also thinking that we may want to upstream this to prompt-toolkit once we battle test it in IPython, so having a generic implementation may be preferable. What do you think?

Reply:
Sure, that sounds reasonable.